dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/DefaultTrackerAccessManager.java

```diff
@@ -107,6 +111,18 @@ public List<String> canWrite(UserDetails user, TrackedEntity trackedEntity) {
   @Override
   public List<String> canRead(
       UserDetails user, TrackedEntity trackedEntity, Program program, boolean skipOwnershipCheck) {
+    List<String> errors = canReadProgramAndTrackedEntityType(user, trackedEntity, program);
+
+    if (!skipOwnershipCheck && !ownershipAccessManager.hasAccess(user, trackedEntity, program)) {
+      errors.add(OWNERSHIP_ACCESS_DENIED);
+    }
+
+    return errors;
+  }
+
+  @Override
+  public List<String> canReadProgramAndTrackedEntityType(
```

Comment:
When would I use this method over `public List<String> canRead(UserDetails user, TrackedEntity trackedEntity, Program program, boolean skipOwnershipCheck)`?

The signatures are so close 😅 

Also if all we do is `!errors.isEmpty()` we should return a bool.

Reply:
I added a javadoc to both methods, hopefully is a bit clearer.
The difference is `canRead` validates the TE/program ownership, while `canReadProgramAndTrackedEntityType` does not.
Anyway, this is needed now because of the way the let the clients know it's possible to break the glass. This is something we need to fix (there's at ticket for it already). I think we should be able to remove this method when we work on it.